waspc/src/Analyzer/TypeChecker/Internal.hs

```diff
@@ -0,0 +1,232 @@
+--   Right (DictType (M.singleton "a" (DictOptional NumberType)))
+unifyTypes :: Type -> Type -> Either TypeError Type
+unifyTypes s t
+  | s == t = Right s
+-- Two lists unify only if their inner types unify
+unifyTypes typS@(ListType s) typT@(ListType t) =
+  fmap ListType $
+    left (\e -> UnificationError (ReasonList e) typS typT) $
+      unifyTypes s t
+-- Declarations and enums can not unify with anything
+unifyTypes s@(DeclType _) t = Left $ UnificationError ReasonDecl s t
+unifyTypes s@(EnumType _) t = Left $ UnificationError ReasonEnum s t
+-- The unification of two dictionaries is defined by the [DictNone] and [DictSome] rules
+unifyTypes typS@(DictType s) typT@(DictType t) = do
+  let keys = M.keysSet s <> M.keysSet t
+  unifiedType <- foldMapM' (\key -> M.singleton key <$> unifyEntryTypesForKey key) keys
+  return $ DictType unifiedType
+  where
+    unifyEntryTypesForKey :: String -> Either TypeError DictEntryType
+    unifyEntryTypesForKey key = annotateError key $ case (M.lookup key s, M.lookup key t) of
+      (Nothing, Nothing) ->
+        error "impossible: unifyTypes.unifyEntryTypesForKey should be called with only the keys of s and t"
+      -- [DictSome] on s, [DictNone] on t
+      (Just sType, Nothing) ->
+        Right $ DictOptional $ dictEntryType sType
+      -- [DictNone] on s, [DictSome] on t
+      (Nothing, Just tType) ->
+        Right $ DictOptional $ dictEntryType tType
+      -- Both require @key@, so it must be a required entry of the unified entry types
+      (Just (DictRequired sType), Just (DictRequired tType)) ->
+        DictRequired <$> unifyTypes sType tType
+      -- One of s or t has @key@ optionally, so it must be an optional entry of the unified entry types
+      (Just sType, Just tType) ->
+        DictOptional <$> unifyTypes (dictEntryType sType) (dictEntryType tType)
+
+    annotateError :: String -> Either TypeError a -> Either TypeError a
+    annotateError k = left (\e -> UnificationError (ReasonDictWrongKeyType k e) typS typT)
+unifyTypes s t = Left $ UnificationError ReasonUncoercable s t
+
+-- | @weaken expr typ == Right expr'@ establishes that @expr@ can be typed as @typ@.
+-- Additionally, the following property is guaranteed:
+--
+-- * @exprType expr == typ@
+--
+-- When a @Left@ value is returned, then @expr@ can not be typed as @typ@.
+weaken :: Type -> TypedExpr -> Either TypeError TypedExpr
+weaken typ' expr
+  | exprType expr == typ' = Right expr
+-- A list can be weakened to @typ@ if
+-- - @typ@ is of the form @ListType typ'@
+-- - Every value in the list can be weakened to @typ'@
+weaken (ListType typ') expr@(List vals _) =
+  fmap (\values -> List values (ListType typ')) $
+    left (\e -> WeakenError (ReasonList e) expr typ') $
+      mapM (weaken typ') vals
+weaken (DictType typ') expr@(Dict entries _) = do
+  entries' <- mapM weakenEntry entries
+  mapM_ guardHasEntry $ M.toList typ'
```

Comment:
I find this line hard to read. Who has entry? What does it mean it has entry?
I know comments are below, but it feels like name could do the explaining instead of me going to understand the comments below.
What makes this harder is that guardHasEntry takes some args, but then also uses some stuff directly from the scope, so it is hard to say what it is operating on exactly.

What does it really do? From what I understood, it check that all required dict entries from the type we want to weaken the expression to are present in the entries of that same expression. So name should reflect that. Maybe smth like `checkThatExprSatisfiesEntryType`?